src/query/functions-v2/src/aggregates/aggregate_window_funnel.rs

```diff
@@ -0,0 +1,450 @@
+use common_expression::Column;
+use common_expression::ColumnBuilder;
+use common_expression::Scalar;
+use common_io::prelude::*;
+use num_traits::AsPrimitive;
+use serde::de::DeserializeOwned;
+use serde::Deserialize;
+use serde::Serialize;
+
+use super::AggregateFunctionRef;
+use super::AggregateNullVariadicAdaptor;
+use super::StateAddr;
+use crate::aggregates::aggregate_function_factory::AggregateFunctionDescription;
+use crate::aggregates::assert_unary_params;
+use crate::aggregates::assert_variadic_arguments;
+use crate::aggregates::AggregateFunction;
+
+#[derive(Serialize, Deserialize)]
+struct AggregateWindowFunnelState<T> {
+    #[serde(bound(deserialize = "T: DeserializeOwned"))]
+    pub events_list: Vec<(T, u8)>,
+    pub sorted: bool,
+}
+
+impl<T> AggregateWindowFunnelState<T>
+where T: Ord
+        + Sub<Output = T>
+        + AsPrimitive<u64>
+        + Serialize
+        + DeserializeOwned
+        + Clone
+        + Send
+        + Sync
+{
+    pub fn new() -> Self {
+        Self {
+            events_list: Vec::new(),
+            sorted: true,
+        }
+    }
+    #[inline(always)]
+    fn add(&mut self, timestamp: T, event: u8) {
+        if self.sorted && !self.events_list.is_empty() {
+            let last = self.events_list.last().unwrap();
+            if last.0 == timestamp {
+                self.sorted = last.1 <= event;
+            } else {
+                self.sorted = last.0 <= timestamp;
+            }
+        }
+        self.events_list.push((timestamp, event));
+    }
+
+    #[inline(always)]
+    fn merge(&mut self, other: &mut Self) {
+        if other.events_list.is_empty() {
+            return;
+        }
+        let l1 = self.events_list.len();
+        let l2 = other.events_list.len();
+
+        self.sort();
+        other.sort();
+        let mut merged = Vec::with_capacity(self.events_list.len() + other.events_list.len());
+        let cmp = |a: &(T, u8), b: &(T, u8)| {
+            let ord = a.0.cmp(&b.0);
+            if ord == Ordering::Equal {
+                a.1.cmp(&b.1)
+            } else {
+                ord
+            }
+        };
+
+        {
+            let mut i = 0;
+            let mut j = 0;
+            let mut k = 0;
+            while i < l1 && j < l2 {
+                if cmp(&self.events_list[i], &other.events_list[j]) == Ordering::Less {
+                    merged.push(self.events_list[i]);
+                    k += 1;
+                    i += 1;
+                } else {
+                    merged.push(other.events_list[j]);
+                    k += 1;
+                    j += 1;
+                }
+            }
+
+            unsafe {
+                merged.set_len(self.events_list.len() + other.events_list.len());
+            }
+
+            if i < l1 {
+                merged[k..].copy_from_slice(&self.events_list[i..]);
+            }
+            if j < l2 {
+                merged[k..].copy_from_slice(&other.events_list[j..]);
```

Comment:
seems this is a bug, `k` should be increased.

Reply:
The value of `k` has already increased in there:

```rust
while i < l1 && j < l2 {
                if cmp(&self.events_list[i], &other.events_list[j]) == Ordering::Less {
                    merged.push(self.events_list[i]);
                    k += 1;
                    i += 1;
                } else {
                    merged.push(other.events_list[j]);
                    k += 1;
                    j += 1;
                }
            }
```

although we increased k, the k will never be read.

Reply:
I think sundy meant after 

```rust
if i < l1 {
    merged.extend(self.events_list.iter());
    merged[k..].copy_from_slice(&self.events_list[i..]);
}
```

`k` should be increased for the next block

```rust
if j < l2 {
    merged.extend(self.events_list.iter());
    merged[k..].copy_from_slice(&self.events_list[j..]);
}
```